Fix filtering azure inventory based on user-specified tags

fosterseth opened the pull request:
Fix filtering azure inventory based on user-specified tags

fix for issue #5044 

Users can specify a list of keys, or a list of key:value pairs under source variables. e.g. `tags: Creator, peanutbutter` or `tags: Creator:jmarshall, peanutbutter:jelly`. If provided, only hosts that have all keys or key:value pairs in the list will be returned.

inventory.py sets up the azure_rm.yml that includes the `exclude_host_filters` field. This code adds a line for each key in the list, as well as an additional line in the case of a key:value pair.

e.g. 

```
exclude_host_filters:
- "'Creator' not in tags.keys()"
- tags['Creator'] not in ['jmarshall']
```

Each line is a conditional, and if any conditionals are true, then the host is filtered out.

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 9.0.1

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

AlanCoding commented:
Could you add a test case that does exactly what you have explained here? At:

https://github.com/ansible/awx/blob/8354137774af11cfc82a8693ca6853274c60d218/awx/main/tests/functional/test_inventory_source_injectors.py#L31-L33

I'm pretty sure I understand the syntax to inputs tags (via source_variables), but it's kind of shaky knowledge so it would be good to offload what you know to the test there. Once you do this, it should start failing, requiring a corresponding change to the data file.

AlanCoding commented on awx/main/models/inventory.py:
```diff
@@ -2036,9 +2036,25 @@ def inventory_as_dict(self, inventory_update, private_data_dir):
+                        user_filters.append('"{}" not in tags.keys()'.format(kv[0].strip()))
+                        # if a value is provided, check that the key:value pair matches
+                        if len(kv) > 1:
+                            user_filters.append('tags["{}"] not in ["{}"]'.format(kv[0].strip(), kv[1].strip()))
```
This `in <a list>` syntax doesn't make sense to me if unless `kv[1]` is split on some delimiter to allow multiple values. If not (which is fine), then I'd rather see this read as `tags["{}"] != "{}"`, as long as that works

fosterseth replied:
yeah this makes sense since we're assuming the value (in the key:value pair) is a single string, not a list.

fosterseth committed:
Fix filtering azure inventory based on user-specified tags

Users can specify a list of keys, or a list of key:value pairs under
source variables. e.g. tags: Creator, peanutbutter or tags:
Creator:jmarshall, peanutbutter:jelly. If provided, only hosts that have
all keys or key:value pairs in the list will be returned.

inventory.py sets up the azure_rm.yml that provides
exclude_host_filters. This code adds a line for each key in the list, as
well as an additional line in the case of a key:value pair.

e.g.

exclude_host_filters:
- "'Creator' not in tags.keys()"
- tags['Creator'] != 'jmarshall'

Each line is a conditional, and if any conditionals is true, then the
host is filtered out.

fix for issue # 5044

AlanCoding reviewed (approved):
perfect, thanks!

fosterseth commented:
recheck
